Archipelago_Gifting.cpp

```
@@ -131,48 +84,70 @@ std::vector<AP_Gift> AP_CheckGifts() {
-        req.operations = {
-            {"pop", &json_string}
-        };
-        AP_Gift cache = cur_gifts_available[giftindex];
-        AP_SetServerData(&req);
-        while (req.status == AP_RequestStatus::Pending && AP_GetConnectionStatus() == AP_ConnectionStatus::Authenticated) {}
-        if (req.status == AP_RequestStatus::Done) {
-            while (findGiftByID(id) != -1) {} // When this value is deleted, we can be sure that the gift is no longer there
-            *gift = cache;
-            return req.status;
-        }
-    }
-    return AP_RequestStatus::Error;
+AP_RequestStatus AP_AcceptGift(std::string id) {
+    return AP_AcceptGift(std::set<std::string>{ id });
+}
+AP_RequestStatus AP_AcceptGift(std::set<std::string> ids) {
+   if (ids.size() == 0)
+        return AP_RequestStatus::Done;
+
+    AP_SetServerDataRequest req;
+    req.key = AP_PLAYER_GIFTBOX_KEY;
+    req.type = AP_DataType::Raw;
+    req.want_reply = true;
+    req.operations = std::vector<AP_DataStorageOperation>();
+
+    std::vector<std::string> id_strings;
+    for (std::string id : ids) 
+        id_strings.push_back("\"" + id + "\"");
+    for (size_t i=0; i < id_strings.size(); i++)
+        req.operations.push_back({"pop", &id_strings[i]});
+    
+    AP_SetServerData(&req);
+    while (req.status == AP_RequestStatus::Pending && AP_GetConnectionStatus() == AP_ConnectionStatus::Authenticated) {}
+    return req.status;
 }
 
 AP_RequestStatus AP_RejectGift(std::string id) {
-    if (findGiftByID(id) != -1) {
-        AP_Gift gift;
-        AP_RequestStatus status = AP_AcceptGift(id, &gift);
-        if (status == AP_RequestStatus::Error) {
-            return status;
+    return AP_RejectGift(std::set<std::string>{ id });
+}
+AP_RequestStatus AP_RejectGift(std::set<std::string> ids) {
+    std::vector<AP_Gift> gifts;
+    std::set<std::string> giftIds;
+    std::vector<AP_Gift> availableGiftsCopy = cur_gifts_available;
+    for (std::string id : ids) {
```

This loop ignores the case where ids given in parameter did not exist. Should one such id exist, it should return an error